httpd/handler.go

```diff
@@ -20,121 +25,84 @@ import (
-	username, password := q.Get("u"), q.Get("p")
-	if username != "" && password != "" {
-		return username, password, nil
-	}
-	auth := r.Header.Get("Authorization")
-	if auth == "" {
-		return "", "", nil
-	}
-	fields := strings.Split(auth, " ")
-	if len(fields) != 2 {
-		return "", "", fmt.Errorf("invalid Basic Authentication header")
-	}
-	bs, err := base64.StdEncoding.DecodeString(fields[1])
-	if err != nil {
-		return "", "", fmt.Errorf("invalid Base64 encoding")
-	}
-	fields = strings.Split(string(bs), ":")
-	if len(fields) != 2 {
-		return "", "", fmt.Errorf("invalid Basic Authentication value")
-	}
-	return fields[0], fields[1], nil
+type route struct {
+	name        string
+	method      string
+	pattern     string
+	handlerFunc func(http.ResponseWriter, *http.Request, *influxdb.User)
 }
 
 // Handler represents an HTTP handler for the InfluxDB server.
 type Handler struct {
-	server *Server
-	mux    *pat.PatternServeMux
-
-	// Whether endpoints require authentication.
-	AuthenticationEnabled bool
-
-	// The InfluxDB verion returned by the HTTP response header.
-	Version string
+	server                *influxdb.Server
+	routes                []route
+	mux                   *pat.PatternServeMux
+	requireAuthentication bool
 }
 
 // NewHandler returns a new instance of Handler.
-func NewHandler(s *Server) *Handler {
+func NewHandler(s *influxdb.Server, requireAuthentication bool, version string) *Handler {
 	h := &Handler{
 		server: s,
 		mux:    pat.New(),
+		requireAuthentication: requireAuthentication,
 	}
 
-	// Query serving route.
-	h.mux.Get("/query", h.makeAuthenticationHandler(h.serveQuery))
-
-	// Data-ingest route.
-	h.mux.Post("/write", h.makeAuthenticationHandler(h.serveWrite))
-
-	// Data node routes.
-	h.mux.Get("/data_nodes", h.makeAuthenticationHandler(h.serveDataNodes))
-	h.mux.Post("/data_nodes", h.makeAuthenticationHandler(h.serveCreateDataNode))
-	h.mux.Del("/data_nodes/:id", h.makeAuthenticationHandler(h.serveDeleteDataNode))
-
-	// Utilities
-	h.mux.Get("/metastore", h.makeAuthenticationHandler(h.serveMetastore))
-	h.mux.Get("/ping", h.makeAuthenticationHandler(h.servePing))
+	weblog := log.New(os.Stderr, `[http] `, 0)
+
+	h.routes = append(h.routes,
+		route{
+			"query", // Query serving route.
+			"GET", "/query", h.serveQuery,
+		},
+		route{
+			"write", // Data-ingest route.
+			"POST", "/write", h.serveWrite,
+		},
+		route{ // List data nodes
+			"data_nodes_index",
+			"GET", "/data_nodes", h.serveDataNodes,
+		},
+		route{ // Create data node
+			"data_nodes_create",
+			"POST", "/data_nodes", h.serveCreateDataNode,
+		},
+		route{ // Delete data node
+			"data_nodes_delete",
+			"DELETE", "/data_nodes/:id", h.serveDeleteDataNode,
+		},
+		route{ // Metastore
+			"metastore",
+			"GET", "/metastore", h.serveMetastore,
+		},
+		route{ // Ping
+			"ping",
+			"GET", "/ping", h.servePing,
+		},
+	)
+
+	for _, r := range h.routes {
+
+		handler := authorize(r.handlerFunc, h, requireAuthentication)
```

Comment:
That's a fair point, and I gave it some thought but didn't implement it.  Two things:  1) Only wrap the code if authorize is on.  2) Don't pass in user to a signature that never uses it.  This might be worth a 10 minute refactor, as I agree that it just makes sense.